File: 2025/day11/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.*;

public class Solution {
    private static Map<String, List<String>> graph;

    public static void main(String[] args) throws IOException {
        String inputFile = "../input.txt";
        if (args.length > 0) {
            inputFile = args[0];
        }

        graph = parseInput(inputFile);

        System.out.println("Part 1: " + part1());
        System.out.println("Part 2: " + part2());
    }

    private static Map<String, List<String>> parseInput(String filename) throws IOException {
        try (Stream<String> lines = Files.lines(Paths.get(filename))) {
            return lines
                .map(String::trim)
                .filter(line -> !line.isEmpty())
                .collect(Collectors.toMap(
                    line -> line.split(": ")[0],
                    line -> {
                        String[] parts = line.split(": ");
                        if (parts.length > 1) {
                            return Arrays.asList(parts[1].split(" "));
                        }
                        return new ArrayList<String>();
                    }
                ));
        }
    }

    private static long part1() {
        Map<String, Long> memo = new HashMap<>();
        return countPaths("you", "out", memo);
    }

    private static long countPaths(String node, String target, Map<String, Long> memo) {
        // Use Objects.hash for efficient memoization key
        String key = node + ":" + target;

        Long cached = memo.get(key);
        if (cached != null) {
            return cached;
        }

        if (node.equals(target)) {
            return 1;
        }

        List<String> neighbors = graph.get(node);
        if (neighbors == null) {
            return 0;
        }

        long total = 0;
        for (String neighbor : neighbors) {
            total += countPaths(neighbor, target, memo);
        }

        memo.put(key, total);
        return total;
    }

    private static long part2() {
        // Create three separate memoization caches for counting paths to different targets
        Map<String, Long> pathsToOut = new HashMap<>();
        Map<String, Long> pathsToDac = new HashMap<>();
        Map<String, Long> pathsToFft = new HashMap<>();

        // Count paths from svr to each intermediate node
        long svrToDac = countPaths("svr", "dac", pathsToDac);
        long svrToFft = countPaths("svr", "fft", pathsToFft);

        // Count paths between intermediate nodes
        long dacToFft = countPaths("dac", "fft", pathsToFft);
        long fftToDac = countPaths("fft", "dac", pathsToDac);

        // Count paths from intermediate nodes to out
        long dacToOut = countPaths("dac", "out", pathsToOut);
        long fftToOut = countPaths("fft", "out", pathsToOut);

        // Paths that visit dac before fft: svr -> dac -> fft -> out
        long dacBeforeFft = svrToDac * dacToFft * fftToOut;

        // Paths that visit fft before dac: svr -> fft -> dac -> out
        long fftBeforeDac = svrToFft * fftToDac * dacToOut;

        return dacBeforeFft + fftBeforeDac;
    }
}

File: 2025/day11/rust/solution.rs
use std::collections::HashMap;
use std::fs;
use std::io::{self, BufRead};

type Graph = HashMap<String, Vec<String>>;
type Cache = HashMap<String, u64>;

fn parse_input(filename: &str) -> io::Result<Graph> {
    let file = fs::File::open(filename)?;
    let reader = io::BufReader::new(file);
    let mut graph = HashMap::new();

    for line in reader.lines() {
        let line = line?;
        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        let parts: Vec<&str> = line.split(": ").collect();
        let node = parts[0].to_string();
        let neighbors = if parts.len() > 1 {
            parts[1].split_whitespace()
                .map(|s| s.to_string())
                .collect()
        } else {
            Vec::new()
        };

        graph.insert(node, neighbors);
    }

    Ok(graph)
}

fn count_paths_to(graph: &Graph, target: &str, cache: &mut Cache, node: &str) -> u64 {
    if node == target {
        return 1;
    }

    if let Some(&cached) = cache.get(node) {
        return cached;
    }

    let count = if let Some(neighbors) = graph.get(node) {
        neighbors.iter()
            .map(|neighbor| count_paths_to(graph, target, cache, neighbor))
            .sum()
    } else {
        0
    };

    cache.insert(node.to_string(), count);
    count
}

fn part1(graph: &Graph) -> u64 {
    let mut cache = HashMap::new();
    count_paths_to(graph, "out", &mut cache, "you")
}

fn part2(graph: &Graph) -> u64 {
    // Count paths from each node to 'out'
    let mut cache_to_out = HashMap::new();
    let paths_to_out = |node: &str, cache: &mut Cache| -> u64 {
        count_paths_to(graph, "out", cache, node)
    };

    // Count paths from each node to 'dac'
    let mut cache_to_dac = HashMap::new();
    let paths_to_dac = |node: &str, cache: &mut Cache| -> u64 {
        count_paths_to(graph, "dac", cache, node)
    };

    // Count paths from each node to 'fft'
    let mut cache_to_fft = HashMap::new();
    let paths_to_fft = |node: &str, cache: &mut Cache| -> u64 {
        count_paths_to(graph, "fft", cache, node)
    };

    // Paths that visit dac before fft: svr -> dac -> fft -> out
    let svr_to_dac = paths_to_dac("svr", &mut cache_to_dac);
    let dac_to_fft = paths_to_fft("dac", &mut cache_to_fft);
    let fft_to_out = paths_to_out("fft", &mut cache_to_out);
    let dac_before_fft = svr_to_dac * dac_to_fft * fft_to_out;

    // Paths that visit fft before dac: svr -> fft -> dac -> out
    let svr_to_fft = paths_to_fft("svr", &mut cache_to_fft);
    let fft_to_dac = paths_to_dac("fft", &mut cache_to_dac);
    let dac_to_out = paths_to_out("dac", &mut cache_to_out);
    let fft_before_dac = svr_to_fft * fft_to_dac * dac_to_out;

    dac_before_fft + fft_before_dac
}

fn main() -> io::Result<()> {
    let input_file = "../input.txt";
    let graph = parse_input(input_file)?;

    println!("Part 1: {}", part1(&graph));
    println!("Part 2: {}", part2(&graph));

    Ok(())
}
